# runCNA.py
A python wrapper for single-cell or bulk sample CNA calling using CBS algorithm.

# Dependencies
Required softwares: Bowtie2, Samtools, Rscript, AnnotSV.

Python3 packages: Seaborn, Pandas, Matplotlib, Numpy.

R packages: DNAcopy.

# Usage Example
## 1. Edit config file
### Example config.json:
    {
        "BowtieRef": "/path/to/ref/hg38/hg38",   #Specify the bowtie2 reference path.
        "GenomeBuild": "GRCh38",   #Reference genome build (Used in AnnotSV, possible values for human genome: ["GRCh38", "GRCh37"]).
        "OutputPath": "/path/to/output/file",   #Specify the output file path.
        "ScriptDir": "/path/to/scripts/file",   #Specify the scripts file path.
        "BinFlag": "200K",   #Specify the bin size. Possible values: ["50K","200K"].
        "Dynamic_Bin": "/path/to/dynamic/bin/file/hg38.varbin.gc.content.200K.bowtie.Umap_k24.NCG.txt",   #Specify the path of dynamic bin file.
        "Bad_Bin": "/path/to/bad/bin/file/hg38.200K.Umap_k24.NCG.bad.txt",   #Specify the path of bad bin file.
        "SD": "1",   #SD value for CBS calling.
        "alpha": "0.0001",   #alpha value for CBS calling.
        "AnnotSVpath": "/path/to/AnnotSV-master/bin/AnnotSV",   #Specify the path of AnnotSV software.
        "CN_Diff_Criterion": "0.5",   #Copy number difference cut-off to identify CNA events. Used in AnnotSV annotation.
        "Samples":    #Sample infomation. 
        {
            #Example sample1.
            "PD10_HYY_20201227M1-10_A1":    #Unique sample ID.
            {
                "seq1": "/path/to/Read1/Fastq/PD10_HYY_20201227M1-10_A1.1.fastq.gz",   #Read1 Fastq path.
                "seq2": "/path/to/Read2/Fastq/PD10_HYY_20201227M1-10_A1.2.fastq.gz",   #Read2 Fastq path.
                "name": "20201227M1-10_A1",   #Sample name.
                "ploidy": "Diploid",   #Sample ploidy. Possible values: ["Diploid","Non-Diploid"].
                "sex": "m"   #Specify sample gender. "m" for male, "f" for female.
            },
            #Example sample2.
            "PD10_HYY_20201227M1-10_A10": 
            {
                "seq1": "/path/to/Read1/Fastq/PD10_HYY_20201227M1-10_A10.1.fastq.gz",
                "seq2": "/path/to/Read2/Fastq/PD10_HYY_20201227M1-10_A10.2.fastq.gz",
                "name": "20201227M1-10_A10",
                "ploidy": "Diploid",
                "sex": "m"
            },
            #Example sample3.
            "PD10_HYY_20201227M1-10_A11": 
            {
                "seq1": "/path/to/Read1/Fastq/PD10_HYY_20201227M1-10_A11.1.fastq.gz",
                "seq2": "/path/to/Read2/Fastq/PD10_HYY_20201227M1-10_A11.2.fastq.gz",
                "name": "20201227M1-10_A11",
                "ploidy": "Diploid",
                "sex": "m"
            }
        }
    }

## 2. Basic calling
### Usage:
    python runCNA.py -c <CONFIG_FILE> -i <SAMPLE_ID>
  The command will map the reads to reference by bowtie2, get the bin counts and perform CNA calling using CBS algorithm. All the sample information and command arguments are stored in the config file.
### Output list:
"Bowtie2.sam": SAM output.

"mapping_log.json": Mapping details.

"CNV.dyna_200K.bin": Original bin count file. For each column in the file, COLUMN 1: Chromosome names; COLUMN 2: Chromosome position of the bin; COLUMN 3: Genome absolute position of the bin; COLUMN 4: Bin counts; COLUMN 5: Normalized bin count ratio.

"CNV.dyna_200K.error": Reads failed in bin assigning.

"SD_1_alpha_0.0001_Copy.200K.txt": CNA profile of the sample. For each column in the file, "chrom": Chromosome names; "chrompos": Chromosome position of the bin; "abspos": Genome absolute position of the bin; "bincount": Bin counts; "ratio": Normalized bin count ratio; "gc.content": GC content of the bin; "lowratio": Bin count ratio after LOWESS normalized; "seg.mean.LOWESS": Segment ratio after LOWESS normalized; "cn.ratio": Final bin values; "cn.seg": Final segment values; "copy.number": Integer copy number values.

"CNAplot.pdf": Whole genome CNA profile of the sample.

## 3. CNA annotation by AnnotSV
### Usage:
    python runCNA.py -m Annotation -c <CONFIG_FILE> -i <SAMPLE_ID> -s <SEGMENT_FILE_NAME>
  The command will annotate the CNAs using AnnotSV software. -s will specify the name of segment file which if ommited, the program will use the default segment file name generated by basic calling step. Please make sure the provided information in the config file matched all the details in each step (e.g. GenomeBuild info is consistent with the genome reference used in bowtie2 mapping step).
### Output list:
"CNAposition.bed": A bed file of all the CNA events required in the annotation. For each column in the file, COLUMN 1: Chromosome names; COLUMN 2: Chromosome position of CNA starts; COLUMN 3: Chromosome position of CNA ends; COLUMN 4: CNA type.

"CNA.annotated.tsv": Annotated CNA tables generated by AnnotSV. For column details please refer to the AnnotSV manual: https://lbgi.fr/AnnotSV/Documentation/README.AnnotSV_latest.pdf.

# Credits
cbs.copy.R was written by Dr. Timour Baslan. uniq2bin.50k.pl was written by Dr. Yusi Fu. runCNA.py was written by Lu Liu inspired by the original scripts written by Dr. Xiannian Zhang.

# Reference
1. Baslan, T., Kendall, J., Rodgers, L., Cox, H., Riggs, M., Stepansky, A., ... & Hicks, J. (2012). Genome-wide copy number analysis of single cells. Nature protocols, 7(6), 1024-1041.
2. Langmead, B., Trapnell, C., Pop, M., & Salzberg, S. L. (2009). Ultrafast and memory-efficient alignment of short DNA sequences to the human genome. Genome biology, 10(3), 1-10.
3. Li, H., Handsaker, B., Wysoker, A., Fennell, T., Ruan, J., Homer, N., ... & Durbin, R. (2009). The sequence alignment/map format and SAMtools. Bioinformatics, 25(16), 2078-2079.
4. Olshen, A. B., Venkatraman, E. S., Lucito, R., & Wigler, M. (2004). Circular binary segmentation for the analysis of array‚Äêbased DNA copy number data. Biostatistics, 5(4), 557-572.
5. Geoffroy, V., Herenger, Y., Kress, A., Stoetzel, C., Piton, A., Dollfus, H., & Muller, J. (2018). AnnotSV: an integrated tool for structural variations annotation. Bioinformatics, 34(20), 3572-3574.
